content/post/performance-metrics-overview.md

```diff
@@ -0,0 +1,111 @@
+
+Here’s an example of state partitioning in Wallaroo:
+
+For a [Word Count](https://docs.wallaroolabs.com/book/python/word-count.html) application, we partition our word count state into distinct state entities based on the letters of the alphabet. We ultimately set up 27 state entities in total, one for each letter plus one called "!" which will handle any "word" that doesn't start with a letter. By partitioning, we remove the potential bottleneck of a single data structure to maintain the state of the count of all of our words.
+
+Each state entity is managed by a Step, an actor which is responsible for running the state computation and routing output messages downstream, amongst other things. Using Steps allow us to avoid coordination in updating state, but we also need to avoid coordination in our metrics capturing system. To get a full picture of how a state computation is performing, the metrics provided by each Step needs to be able to be stitched together. Here's a diagram to best illustrate the type of aggregation we require across Steps:
+
+![aggregated steps diagram](/images/post/performance-metrics-overview/steps-diagram.png)
+
+A [Sorted Values List](https://blog.wallaroolabs.com/2018/02/missing-slug#using-a-sorted-values-list) would give us the granularity we need to do the aggregations required but would have a much higher performance impact than we want. Another option would be to store specific statistics (mean/median/average) per Step, but the need for aggregation would render these stats entirely useless if we want an accurate depiction of the computation multiple Steps represent. [Theo Schlossnagle](https://twitter.com/postwait) wrote a great blog post explaining why bad math like an average of averages provides useless metrics: ["The Problem with Math: Why Your Monitoring Solution is Wrong"](https://www.circonus.com/2015/02/problem-math/). Existing industry research shows that histograms are excellent solutions for gathering meaningful metrics: ["How NOT to Measure Latency"](https://www.youtube.com/watch?v=lJ8ydIuPFeU) by [Gil Tene](https://twitter.com/giltene) and ["The Uphill Battle for Visibility"](https://www.circonus.com/2016/07/uphill-battle-visibility/) by [Theo Schlossnagle](https://twitter.com/postwait) are just two great examples. We recognized that histograms would work well for us due to the following:
```

Comment:
The use of `stats` is unexpected since we've been using the full word `statistics` so far. Maybe change `stats` to `statistics`?